### DISCUSSION TITLE
Data Received Problem

### DISCUSSION THREAD

**Discussion by zdh3303** [microsoft/msquic#2037] [Q&A] [answered]
hello, If the sender sends a large data packet, does the receiver need to  splice data packets received by myself?

**Comment by nibanks** [comment, marked as answer]
QUIC fragments stream data sent by an application into chunks that can fit into UDP datagrams. The peer QUIC stack then reassembles this data before delivering to the application. So, bottom line, the app is free to send a 1 GB buffer on a stream and the QUIC layer will handle breaking it up and reassembling as necessary.

**Reply by zdh3303** [reply]
the sender sends 10000byte buffer，however the receriver receives 5 packets，demo and receiver info are posted below
![image](https://user-images.githubusercontent.com/46858562/135200839-a5c53a44-e19e-415f-941d-de528fb65256.png)
![image](https://user-images.githubusercontent.com/46858562/135200851-237941cd-ca75-4760-99f6-3d395a4ac531.png)

**Reply by thhous-msft** [reply]
Ah. Each receive will be in order, but there is no guarantee that a single send on 1 end will result in a single receive on the other end. It could be broken up into 2 receive callbacks. This is similar to TCP.

**Reply by zdh3303** [reply]
ok，thanks for your reply, I have another question, I want to send n packets continuously, I create one stream and apply these packet on the stream,  so how can I separate the packets on the server side?

**Reply by nibanks** [reply]
You need to have some way to identify "packet" boundaries yourself. For instance, the first 2 bytes of each "packet" are the length of data that follows, so you know where this packet ends and he next one starts.

**Reply by zdh3303** [reply]
thanks (^ ^)

**Reply by KevF85** [reply]
> the sender sends 10000byte buffer，however the receriver receives 5 packets，demo and receiver info are posted below ![image](https://user-images.githubusercontent.com/46858562/135200839-a5c53a44-e19e-415f-941d-de528fb65256.png) ![image](https://user-images.githubusercontent.com/46858562/135200851-237941cd-ca75-4760-99f6-3d395a4ac531.png)

Hey ! I'm trying to feed the buffers to be sent in the client-server sample. How did you managed to create several buffers and feed them ? Thank you
